### DISCUSSION TITLE
Writing a likelihood for a count model to model abundance (as well as distribution)?

### DISCUSSION THREAD

**Discussion by scarlett-taylor** [inlabru-org/inlabru#113] [Q&A]
Hello! 

I recently did a workshop with Joe Watson on implementing a general framework for estimating the 
spatio-temporal distribution of a species using multiple data types using the inlabru package. 

The link to the workshop materials are here: https://github.com/joenomiddlename/DFO_SDM_Workshop_2020 and the Watson et al. 2020 manuscript that this workshop is based around (Estimating animal utilization distributions from multiple data types: a joint spatio-temporal point process framework) can be accessed at this link: https://arxiv.org/pdf/1911.00151.pdf 

The workshop combines opportunistic and survey data within a model to model distribution, and **includes a note that we "can use this method to obtain estimates of population size, but _need to model the counts_ as well as the detections".** 
I talked to Joe about this and he said all I would need is to create a new like() object, define the formula (log-mean of poison count and own intercept; maybe own spatial field and/or covariates), include in with bru() and then we could use predict () to obtain estimates of population size. 

My background in deriving equations is not strong, so I'm wondering if you would have any suggestions of a start point for the likelihood object (i.e. like())? 

For the abundance estimate, I'm strictly using line-transect survey data from two years, 2007 and 2016. From running simple distance sampling analyses with ds() from the R 'Distance' package, I have found that the best model for my 2007 data is a half-normal detection model without covariates, whereas the 2016 data is best modelled by a hazard-rate detection function with multiple (3) covariates. 

Let me know if there's anything else you need to know, or if you have any ideas of how I would start writing the likelihood for the count model. 

Thanks for your time, 
**Scarlett Taylor** 
MSc. Candidate
Integrated Fisheries Lab | Department of Biology | LSC 7088
Dalhousie University | Halifax, NS B3H 4R2

Halifax Coordinator: 
Sharks of the Atlantic Research and Conservation Centre (ShARCC)

e: scarlett.taylor@dal.ca
c: (905) 872-8534

**Comment by joenomiddlename** [comment]
Hi Scarlett!

I'm glad to see you're still looking to use inlabru! :) 

I'll just summarize the types of models that we fit in the workshops to help the other inlabru collaborators understand the problem. We fit log-Gaussian Cox process models to sightings data that were collected following distance sampling protocols. We assumed that each encounter was with a single individual and we completely ignored group size. We used a half-normal detection function. I believe the issue at hand is how to model both the locations of **groups** and the **group size**. I suggested that a spatially-varying group size distribution be considered (following the suggestion seen in https://besjournals.onlinelibrary.wiley.com/doi/epdf/10.1111/2041-210X.13168) and jointly modelled. This should allow for both the group locations and group sizes to be jointly modelled in a marked LGCP framework.

With regards to deriving the group size like() object, suppose your SpatialPointsDataFrame object is called *data_sp* and only contains a single row per group encounter (e.g. an encounter with a group of size 3 only has a single row of data instead of 3 rows). Next, suppose the group size corresponding to each group encounter is stored in *data_sp* as the variable GROUP_SIZE. Assuming that you have already created the lgcp like() object for the encounters (as was seen in the workshops), then, the following Poisson like() object could be created:

gp_like <- like(
data = data_sp,
formula = GROUP_SIZE ~ gp_Intercept + gp_mySpatial,
family='poisson'
)

The 'poisson' family has a default log link as so this model would read (with mu(**s**) equal to the mean of the group size distribution at location s):

log(mu(**s**)) =  gp_Intercept + gp_mySpatial(**s**)

Now, all that is required is that the two new components 'gp_Intercept' and 'gp_mySpatial' are defined in the components step and the gp_like is added to the bru() call alongside the lgcp like() object (defined in the workshops with either or log_hn or log_hr detection probability functions defined). Note that 'gp_mySpatial' is a Gaussian random field as seen in the workshops. 

This model will hopefully be a useful first step. It basically allows the average group size to vary across space independently from the lgcp intensity surface (described as the encounter rate in the workshops). Covariates and shared terms could be considered in the future.

The next task is to predict the population size given this joint model. This would require sampling both encounter locations (from the fitted lgcp) and multiplying by sampled group sizes (from the fitted poisson model evaluated at the sampled encounter locations). This is probably the part where @finnlindgren will suggest a better approach, but I think the process could look something like this:

Step 1: define the integration points and weights (as described in the workshops). Let 'study_region_sp' be the SpatialPolygonsDataFrame defining the study region and 'mesh' define the triangulation mesh you defined earlier.

predpts <- ipoints(study_region_sp, mesh)

Step 2: sample the mean number of encounters that occur around each integration point, scale by a sample of group size per encounter, and sum over the integration points. Let 'fit' be the fitted bru model object. Then the sampled population size would look something like:

Lambda <- predict(fit, predpts, ~ sum(weight * exp(lgcp_terms) * exp(gp_terms)), n.samples = 1000)

Suppose the lgcp model only contains an intercept and Gaussian random field named 'Intercept' and 'mySpatial' respectively (i.e. lgcp_terms = Intercept + mySpatial), then the above would read:

Lambda <- predict(fit, predpts, ~ sum(weight * exp(Intercept + mySpatial) * exp(gp_Intercept + gp_mySpatial)), n.samples = 1000)

I hope this helps! I may have got a few things wrong in this explanation, so I would suggest waiting for @finnlindgren to have the chance to catch any gross errors :)

**Reply by finnlindgren** [reply]
Yes, @joenomiddlename covered most of it. This modelling approach works as long as the detection model doesn't depend on the group size. We're working on a paper on how to deal with that situation, and I can't share that yet, but can say a bit more about how to handle predict() in the easier case that @joenomiddlename  described.

The posterior expectation of the total count can be estimated as described at the end of Joe's reply. The full posterior distribution would likely involve sampling almost the full process; sampling the intensity fields with the generate() method, then conditionally on those samples generate a sample of the point process itself, and then sample from the group size distribution at the generated points. But if only the mean and variance are needed, then predict() is sufficient. When the group size is modelled by a Poisson (which can't be entirely correct, since that allows the group size to be zero!), we can work out what the posterior expectation and variances are:

Let $\lambda_n(s)$ be the point process intensity, and let $Z(s)$ be the potential group size at location $s$, with expectation $\lambda_z(s)$, and let $T=\sum_i Z(s_i)$, where $s_i$ are the actual point/group locations. Then we get the conditional expectation as $m=E(T|\lambda))=\int \lambda_n(s) \lambda_z(s) ds$, and the conditional variance as $v=Var(T|\lambda)=\int \lambda_n(s) \lambda_z(s)^2 ds$. In the posterior, we get expectation $E(T)=E(E(T|\lambda))=E(m)$, and variance $Var(T)=E(Var(T|\lambda))+Var(E(T|\lambda))=E(v)+Var(e)$. This shows that what Joe showed with predict() for the posterior expectation is correct. It also shows that to get the posterior variance, we need to take the sum of the posterior variance obtained from that predict call and add it to the posterior expectation from
```
Lambda_var <- predict(
  fit,
  predpts,
  ~ sum(weight * exp(Intercept + mySpatial) * exp(gp_Intercept + gp_mySpatial)^2 ),
  n.samples = 1000)
```
This can all be combined into a single predict() call (which will be almost twice as fast as it needs only a single internal call to `generate()`):
```
Lambda_post <- predict(
  fit,
  predpts,
  ~ c(
       expect = sum(weight * exp(Intercept + mySpatial) * exp(gp_Intercept + gp_mySpatial)),
       variance = sum(weight * exp(Intercept + mySpatial) * exp(gp_Intercept + gp_mySpatial)^2 )
     ),
  n.samples = 1000)
# The rownames of the output should now be "Lambda_expect" and "Lambda_variance"
count_expectation <- Lambda_post["expect", "mean"]
count_variance <- Lambda_post["variance", "mean"] + Lambda_post["expect", "sd"]^2
```
At least this is how it should work in theory; I haven't tested the code for this particular model.
For other distribution models for the group size, the calculations for the posterior variance would need to change a bit, since the conditional variance would no longer be equal to the conditional expectation.

**Reply by scarlett-taylor** [reply]
Hello @joenomiddlename and @finnlindgren, 

Thanks for such a quick reply! I appreciate the help. I'm lucky in the sense that I'm modelling basking sharks - which are relatively solitary (or at least they are in my data!) - so group size is not a large concern. I'll try this all out and get back to you with an update relatively soon. 

Cheers!
